src/main/scala/is/hail/asm4s/package.scala

```diff
@@ -25,17 +28,19 @@ package object asm4s {
     val slots: Int = 1
 
     def newArray(): AbstractInsnNode
+    def classTag: ClassTag[T] = reflect.classTag
   }
 
-  sealed trait MaybeGenericTypeInfo[T] extends TypeInfo[T] {
+  sealed abstract class MaybeGenericTypeInfo[T : TypeInfo] extends TypeInfo[T]()(typeInfo[T].classTag) {
```

Comment:
You're implementing a TypeInfo for `T` but require it already have `TypeInfo`?  That seems wrong.  As above, just require `T: ClassTag` and then ...

Reply:
The existing implementation also requires a `TypeInfo` (albeit not on the superclass) because the two instantiations wrap around a "primitive" `TypeInfo` to relate that primitive type to a boxed type, for Generics.

I could create a separate hierarchy:

```scala
sealed abstract class GenericTypeInfo[T: TypeInfo] {
  def castFromGeneric: Code[_] => Code[T]
  def castToGeneric: Code[T] => Code[_]
  val concrete: TypeInfo[_]
  val generic: TypeInfo[_]
  val isGeneric: Boolean
}
```

I didn't do this because I wanted to use a "generic" `TypeInfo` everywhere I could use a non-generic one, but maybe that's not a good idea.

---

These classes are only used in the `FunctionBuilder` to correctly implement a generic interface. This is related to what we discussed the other day. For `FunctionBuilder` I fully specify the instantiation of the possibly generic interface that the FB is implementing.